File: frontend-search/src/lib/api/kratos.ts
/**
 * Kratos API Helper Library
 *
 * This module provides helper functions for interacting with Ory Kratos
 * authentication flows via our backend proxy endpoints.
 */

import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_URL || 'https://api.arack.io';
const KRATOS_PUBLIC_URL = 'http://127.0.0.1:4433';

/**
 * Kratos Flow Response
 */
export interface KratosFlow {
	id: string;
	type?: string;
	expires_at?: string;
	issued_at?: string;
	request_url?: string;
	ui: {
		action: string;
		method: string;
		nodes: Array<{
			type: string;
			group: string;
			attributes: {
				name: string;
				type?: string;
				value?: any;
				required?: boolean;
				disabled?: boolean;
			};
			messages?: Array<{
				id: number;
				text: string;
				type: string;
			}>;
		}>;
		messages?: Array<{
			id: number;
			text: string;
			type: string;
		}>;
	};
}

/**
 * Kratos Session Response
 */
export interface KratosSession {
	id: string;
	email: string;
	first_name?: string;
	last_name?: string;
	authenticated: boolean;
}

/**
 * Registration Data
 */
export interface RegistrationData {
	email: string;
	password: string;
	first_name: string;
	last_name: string;
	username: string;
	date_of_birth: string;
	gender: string;
}

/**
 * Login Data
 */
export interface LoginData {
	identifier: string;
	password: string;
}

/**
 * Initialize registration flow
 *
 * This creates a new registration flow directly from Kratos and returns the complete flow
 * including the CSRF token and action URL.
 */
export async function initRegistrationFlow(): Promise<KratosFlow> {
	try {
		// Call Kratos directly (not through backend proxy) to get proper CSRF token
		const response = await axios.get('https://auth.arack.io/self-service/registration/api', {
			withCredentials: true // Important: Get CSRF cookie
		});
		return response.data;
	} catch (error: any) {
		console.error('Failed to initialize registration flow:', error);
		throw new Error(error.response?.data?.error || 'Failed to initialize registration');
	}
}

/**
 * Submit registration data
 *
 * Submits the registration form directly to Kratos using the flow's action URL.
 * On success, the user is registered and logged in.
 *
 * Following Ory best practices: accepts complete flow object (no second fetch)
 */
export async function submitRegistration(
	flow: KratosFlow,
	data: RegistrationData
): Promise<void> {
	try {
		// Extract CSRF token from the flow object
		const csrfToken = getCsrfToken(flow);
		if (!csrfToken) {
			throw new Error('CSRF token not found in flow');
		}

		// Submit directly to Kratos using the flow's action URL
		const response = await axios.post(
			flow.ui.action,
			{
				csrf_token: csrfToken,
				method: 'password',
				'traits.email': data.email,
				password: data.password,
				'traits.first_name': data.first_name,
				'traits.last_name': data.last_name,
				'traits.username': data.username,
				'traits.date_of_birth': data.date_of_birth,
				'traits.gender': data.gender
			},
			{
				headers: {
					'Content-Type': 'application/json',
					Accept: 'application/json'
				},
				withCredentials: true // Important: Include CSRF cookie
			}
		);

		// Registration successful - session cookie is automatically set
		return response.data;
	} catch (error: any) {
		console.error('Registration failed:', error);

		// Extract error message from Kratos response
		const kratosError = error.response?.data?.error;
		if (kratosError) {
			throw new Error(`Registration failed: ${JSON.stringify(kratosError)}`);
		}

		throw new Error(error.response?.data?.message || 'Registration failed');
	}
}

/**
 * Initialize login flow
 *
 * This creates a new login flow directly from Kratos and returns the complete flow
 * including the CSRF token and action URL.
 */
export async function initLoginFlow(): Promise<KratosFlow> {
	try {
		// Call Kratos directly (not through backend proxy) to get proper CSRF token
		const response = await axios.get('https://auth.arack.io/self-service/login/api', {
			withCredentials: true // Important: Get CSRF cookie
		});
		return response.data;
	} catch (error: any) {
		console.error('Failed to initialize login flow:', error);
		throw new Error(error.response?.data?.error || 'Failed to initialize login');
	}
}

/**
 * Submit login credentials
 *
 * Submits the login form directly to Kratos using the flow's action URL.
 * On success, a session cookie is set.
 *
 * Following Ory best practices: accepts complete flow object (no second fetch)
 */
export async function submitLogin(flow: KratosFlow, data: LoginData): Promise<void> {
	try {
		// Extract CSRF token from the flow object
		const csrfToken = getCsrfToken(flow);
		if (!csrfToken) {
			throw new Error('CSRF token not found in flow');
		}

		// Submit directly to Kratos using the flow's action URL
		const response = await axios.post(
			flow.ui.action,
			{
				csrf_token: csrfToken,
				method: 'password',
				identifier: data.identifier,
				password: data.password
			},
			{
				headers: {
					'Content-Type': 'application/json',
					Accept: 'application/json'
				},
				withCredentials: true // Important: Include CSRF cookie
			}
		);

		// Login successful - session cookie is automatically set
		return response.data;
	} catch (error: any) {
		console.error('Login failed:', error);

		// Extract error message from Kratos response
		const kratosError = error.response?.data?.error;
		if (kratosError) {
			throw new Error(`Login failed: ${JSON.stringify(kratosError)}`);
		}

		throw new Error(error.response?.data?.message || 'Invalid credentials');
	}
}

/**
 * Get current session (whoami)
 *
 * Checks if the user is authenticated by validating the session cookie.
 * Returns user information if authenticated, throws error otherwise.
 */
export async function whoami(): Promise<KratosSession> {
	try {
		// Call Kratos directly to check session
		const response = await axios.get('https://auth.arack.io/sessions/whoami', {
			withCredentials: true // Important: Send session cookie
		});

		// Kratos returns session data directly
		const session = response.data;

		// Extract user information from Kratos identity
		const identity = session.identity;
		const traits = identity?.traits || {};

		return {
			id: identity?.id || '',
			email: traits.email || '',
			first_name: traits.first_name || traits.name?.first || '',
			last_name: traits.last_name || traits.name?.last || '',
			authenticated: session.active || false
		};
	} catch (error: any) {
		// 401 means not authenticated (expected for logged-out users)
		if (error.response?.status === 401) {
			throw new Error('Not authenticated');
		}

		console.error('Whoami failed:', error);
		throw new Error(error.response?.data?.error?.message || 'Not authenticated');
	}
}

/**
 * Logout current user
 *
 * Initiates the logout flow and redirects to complete logout.
 * Clears the session cookie.
 */
export async function logout(): Promise<void> {
	try {
		// Create logout flow directly from Kratos
		const response = await axios.get('https://auth.arack.io/self-service/logout/browser', {
			withCredentials: true
		});

		const logoutToken = response.data.logout_token;

		// Execute logout using the token
		await axios.get(`https://auth.arack.io/self-service/logout?token=${logoutToken}`, {
			withCredentials: true
		});

		// Logout successful - session cookie is cleared
	} catch (error: any) {
		console.error('Logout failed:', error);
		throw new Error(error.response?.data?.error?.message || 'Logout failed');
	}
}

/**
 * Check if flow has expired
 *
 * Kratos flows have a 10-minute lifetime. This helper checks if a flow
 * has expired and needs to be reinitialized.
 */
export function isFlowExpired(flow: KratosFlow): boolean {
	if (!flow.expires_at) return false;

	const expiresAt = new Date(flow.expires_at);
	const now = new Date();

	return now >= expiresAt;
}

/**
 * Extract CSRF token from flow
 *
 * Helper to extract the CSRF token from a Kratos flow.
 * This is needed for form submission.
 */
export function getCsrfToken(flow: KratosFlow): string | null {
	const csrfNode = flow.ui.nodes.find(
		(node) => node.attributes.name === 'csrf_token'
	);

	return csrfNode?.attributes.value || null;
}

/**
 * Extract error messages from flow
 *
 * Helper to extract user-friendly error messages from a Kratos flow response.
 */
export function getFlowErrors(flow: KratosFlow): string[] {
	const errors: string[] = [];

	// Global messages
	if (flow.ui.messages) {
		errors.push(...flow.ui.messages.map((msg) => msg.text));
	}

	// Field-specific messages
	flow.ui.nodes.forEach((node) => {
		if (node.messages) {
			errors.push(...node.messages.map((msg) => msg.text));
		}
	});

	return errors;
}

File: src/lib/ory/kratos.rs
// Phase 8.6: Kratos API Client
// This module provides a client for interacting with Ory Kratos APIs

use anyhow::Result;
use reqwest::Client;
use super::models::{KratosSession, KratosIdentity};

/// Kratos API client
#[derive(Clone)]
pub struct KratosClient {
    client: Client,
    public_url: String,
    admin_url: String,
}

impl KratosClient {
    /// Create a new Kratos client
    pub fn new(public_url: String, admin_url: String) -> Self {
        Self {
            client: Client::new(),
            public_url,
            admin_url,
        }
    }

    /// Validate session using whoami endpoint
    ///
    /// This endpoint checks if a session cookie is valid and returns the session details
    /// including the identity information.
    ///
    /// For API flows, the session token is stored in a cookie named `ory_kratos_session`,
    /// but Kratos expects it to be sent via X-Session-Token header.
    pub async fn whoami(&self, cookie_header: &str) -> Result<KratosSession> {
        // Extract session token from cookie header
        // Format: "ory_kratos_session=ory_st_...; other_cookie=value"
        let session_token = cookie_header
            .split(';')
            .find_map(|cookie| {
                let cookie = cookie.trim();
                if cookie.starts_with("ory_kratos_session=") {
                    Some(cookie.trim_start_matches("ory_kratos_session="))
                } else {
                    None
                }
            })
            .ok_or_else(|| anyhow::anyhow!("No ory_kratos_session cookie found"))?;

        let response = self
            .client
            .get(format!("{}/sessions/whoami", self.public_url))
            .header("X-Session-Token", session_token)
            .send()
            .await?;

        if !response.status().is_success() {
            anyhow::bail!("Whoami request failed: {}", response.status());
        }

        let session: KratosSession = response.json().await?;
        Ok(session)
    }

    /// Get identity by ID (admin API)
    ///
    /// This is an admin endpoint that can fetch any identity by ID.
    /// Useful for debugging or admin operations.
    pub async fn get_identity(&self, identity_id: &uuid::Uuid) -> Result<KratosIdentity> {
        let response = self
            .client
            .get(format!("{}/admin/identities/{}", self.admin_url, identity_id))
            .send()
            .await?;

        if !response.status().is_success() {
            anyhow::bail!("Get identity failed: {}", response.status());
        }

        let identity: KratosIdentity = response.json().await?;
        Ok(identity)
    }

    /// Check if Kratos is healthy (for startup checks)
    pub async fn health_check(&self) -> Result<bool> {
        let response = self
            .client
            .get(format!("{}/health/ready", self.admin_url))
            .send()
            .await?;

        Ok(response.status().is_success())
    }

    /// Exchange session token for session cookies
    ///
    /// Calls whoami with X-Session-Token header to get Set-Cookie headers
    /// that can be forwarded to the browser
    pub async fn exchange_token_for_cookies(&self, session_token: &str) -> Result<Vec<String>> {
        let response = self
            .client
            .get(format!("{}/sessions/whoami", self.public_url))
            .header("X-Session-Token", session_token)
            .send()
            .await?;

        if !response.status().is_success() {
            anyhow::bail!("Token exchange failed: {}", response.status());
        }

        // Extract Set-Cookie headers
        let cookies: Vec<String> = response.headers()
            .get_all("set-cookie")
            .iter()
            .filter_map(|h| h.to_str().ok())
            .map(|s| s.to_string())
            .collect();

        Ok(cookies)
    }

    /// Initialize registration flow
    ///
    /// Returns a registration flow that can be used to submit registration data.
    /// The flow contains UI nodes with CSRF tokens and other required fields.
    pub async fn init_registration_flow(&self) -> Result<serde_json::Value> {
        // Use API flow for backend proxy compatibility
        let url = format!("{}/self-service/registration/api", self.public_url);

        let response = self.client
            .get(&url)
            .send()
            .await?;

        if !response.status().is_success() {
            anyhow::bail!("Registration flow init failed: {}", response.status());
        }

        let flow = response.json().await?;
        Ok(flow)
    }

    /// Initialize login flow
    ///
    /// Returns a login flow that can be used to submit login credentials.
    /// The flow contains UI nodes with CSRF tokens and other required fields.
    /// Uses browser flow to enable cookie-based sessions.
    pub async fn init_login_flow(&self) -> Result<serde_json::Value> {
        // Browser flows don't work well with backend proxies.
        // Use API flow but we'll handle cookies manually
        let url = format!("{}/self-service/login/api", self.public_url);

        let response = self.client
            .get(&url)
            .send()
            .await?;

        if !response.status().is_success() {
            anyhow::bail!("Login flow init failed: {}", response.status());
        }

        let flow = response.json().await?;
        Ok(flow)
    }

    /// Initialize logout flow
    ///
    /// Returns a logout URL that can be used to complete the logout process.
    /// The cookie parameter should be the full Cookie header value from the request.
    pub async fn init_logout_flow(&self, cookie: Option<&str>) -> Result<String> {
        let url = format!("{}/self-service/logout/browser", self.public_url);

        let mut request = self.client.get(&url);

        if let Some(cookie_header) = cookie {
            request = request.header("Cookie", cookie_header);
        }

        let response = request
            .send()
            .await?;

        if !response.status().is_success() {
            anyhow::bail!("Logout flow init failed: {}", response.status());
        }

        let logout_data: serde_json::Value = response.json().await?;

        let logout_url = logout_data["logout_url"]
            .as_str()
            .ok_or_else(|| anyhow::anyhow!("No logout_url in response"))?
            .to_string();

        Ok(logout_url)
    }

    /// Submit registration data
    ///
    /// Submits registration form to Kratos. Returns session cookies in response.
    pub async fn submit_registration(
        &self,
        flow_id: &str,
        email: &str,
        password: &str,
        first_name: &str,
        last_name: &str,
        username: &str,
        date_of_birth: &str,
        gender: &str,
    ) -> Result<(serde_json::Value, Vec<String>)> {
        let url = format!("{}/self-service/registration?flow={}", self.public_url, flow_id);

        let response = self.client
            .post(&url)
            .json(&serde_json::json!({
                "method": "password",
                "traits": {
                    "email": email,
                    "first_name": first_name,
                    "last_name": last_name,
                    "username": username,
                    "date_of_birth": date_of_birth,
                    "gender": gender
                },
                "password": password
            }))
            .send()
            .await?;

        // Extract Set-Cookie headers as separate strings
        let cookies: Vec<String> = response.headers()
            .get_all("set-cookie")
            .iter()
            .filter_map(|h| h.to_str().ok())
            .map(|s| s.to_string())
            .collect();

        if !response.status().is_success() {
            let error_body = response.text().await?;
            anyhow::bail!("Registration failed: {}", error_body);
        }

        let body = response.json().await?;
        Ok((body, cookies))
    }

    /// Submit login credentials
    ///
    /// Submits login form to Kratos. Returns session cookies in response.
    pub async fn submit_login(
        &self,
        flow_id: &str,
        identifier: &str,
        password: &str,
    ) -> Result<(serde_json::Value, Vec<String>)> {
        let url = format!("{}/self-service/login?flow={}", self.public_url, flow_id);

        let response = self.client
            .post(&url)
            .json(&serde_json::json!({
                "method": "password",
                "identifier": identifier,
                "password": password
            }))
            .send()
            .await?;

        // Extract Set-Cookie headers as separate strings
        let cookies: Vec<String> = response.headers()
            .get_all("set-cookie")
            .iter()
            .filter_map(|h| h.to_str().ok())
            .map(|s| s.to_string())
            .collect();

        if !response.status().is_success() {
            let error_body = response.text().await?;
            anyhow::bail!("Login failed: {}", error_body);
        }

        let body = response.json().await?;
        Ok((body, cookies))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_kratos_client_creation() {
        let client = KratosClient::new(
            "http://127.0.0.1:4433".to_string(),
            "http://127.0.0.1:4434".to_string(),
        );
        assert_eq!(client.public_url, "http://127.0.0.1:4433");
        assert_eq!(client.admin_url, "http://127.0.0.1:4434");
    }
}
